nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java

```diff
@@ -0,0 +1,436 @@
+            "  String test(@Nullable Object o1) {",
+            "    if (Validation.identity(null == o1)) {",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      return o1.toString();",
+            "    } else {",
+            "      return o1.toString();",
+            "    }",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void invertNotNull() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    if (Validation.invert(null != o1)) {",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      return o1.toString();",
+            "    } else {",
+            "      return o1.toString();",
+            "    }",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void checkAndReturn() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "import org.jetbrains.annotations.Contract;",
+            "class Test {",
+            "  @Contract(\"false -> fail\")",
+            "  static boolean checkAndReturn(boolean value) {",
+            "    if (!value) {",
+            "      throw new RuntimeException();",
+            "    }",
+            "    return true;",
+            "  }",
+            "  String test1(@Nullable Object o1, @Nullable Object o2) {",
+            "    if (checkAndReturn(o1 != null) && o2 != null) {",
+            "      return o1.toString() + o2.toString();",
+            "    } else {",
+            "      return o1.toString() + ",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "        o2.toString();",
+            "    }",
+            "  }",
+            "  boolean test2(@Nullable Object o1, @Nullable Object o2) {",
+            "    return checkAndReturn(o1 != null) && o1.toString().isEmpty();",
+            "  }",
+            "  boolean test3(@Nullable Object o1, @Nullable Object o2) {",
+            "    return checkAndReturn(o1 == null) ",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      && o1.toString().isEmpty();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void contractUnreachablePath() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "class Test {",
+            "  String test(Object required) {",
+            "    return Validation.identity(required == null)",
+            "      ? required.toString()",
+            "      : required.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void contractUnreachablePathAfterFailure() {
```

Comment:
What is this test testing that `identityIsNull()` does not? Is this meant to be:

```
            "    Validation.checkFalse(o == null);",
            "    return Validation.identity(o == null)",
            "      // No error because the then-branch is unreachable",
            "      ? o.toString()",
            "      : o.toString();",
```

(Whether or not that is actually handled or a known FP)?

Reply:
I should have done a better job adding comments inline with these tests. This is meant to ensure the `checkTrue(o == null)` refines `o` to `Nullness.NULL` and correctly flags that the is-null case will NPE on `o.toString()` while nothing is erroneously reported in the unreachable else clause.